cpp/src/arrow/array/builder_run_end.cc

```diff
@@ -0,0 +1,331 @@
+  ARROW_DCHECK(array.type->Equals(type_));
+
+  // Ensure any open run is closed before appending the array slice.
+  RETURN_NOT_OK(value_run_builder_->FinishCurrentRun());
+
+  if (length == 0) {
+    return Status::OK();
+  }
+
+  ArraySpan to_append = array;
+  to_append.SetSlice(array.offset + offset, length);
+
+  switch (type_->run_end_type()->id()) {
+    case Type::INT16:
+      RETURN_NOT_OK(DoAppendArray<int16_t>(to_append));
+      break;
+    case Type::INT32:
+      RETURN_NOT_OK(DoAppendArray<int32_t>(to_append));
+      break;
+    case Type::INT64:
+      RETURN_NOT_OK(DoAppendArray<int64_t>(to_append));
+      break;
+    default:
+      return Status::Invalid("Invalid type for run ends array: ", type_->run_end_type());
+  }
+
+  return Status::OK();
+}
+
+std::shared_ptr<DataType> RunEndEncodedBuilder::type() const { return type_; }
+
+Status RunEndEncodedBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
+  // Finish the values array before so we can close the current run and append
+  // the last run-end.
+  std::shared_ptr<ArrayData> values_data;
+  RETURN_NOT_OK(value_run_builder_->FinishInternal(&values_data));
+  auto values_array = MakeArray(values_data);
+
+  ARROW_ASSIGN_OR_RAISE(auto run_ends_array, run_end_builder().Finish());
+
+  ARROW_ASSIGN_OR_RAISE(auto ree_array,
+                        RunEndEncodedArray::Make(length_, run_ends_array, values_array));
+  *out = std::move(ree_array->data());
+  return Status::OK();
+}
+
+Status RunEndEncodedBuilder::FinishCurrentRun() {
+  RETURN_NOT_OK(value_run_builder_->FinishCurrentRun());
+  UpdateDimensions(length_, 0);
+  return Status::OK();
+}
+
+template <typename RunEndsType>
+Status RunEndEncodedBuilder::DoAppendRunEnd(int64_t run_end) {
+  constexpr auto max = std::numeric_limits<RunEndsType>::max();
+  if (ARROW_PREDICT_FALSE(run_end > max)) {
+    return Status::Invalid("Run end value must fit on run ends type but ", run_end, " > ",
+                           max, ".");
+  }
+  return internal::checked_cast<typename CTypeTraits<RunEndsType>::BuilderType*>(
+             children_[0].get())
+      ->Append(static_cast<RunEndsType>(run_end));
+}
+
+Status RunEndEncodedBuilder::AppendRunEnd(int64_t run_end) {
+  switch (type_->run_end_type()->id()) {
+    case Type::INT16:
+      RETURN_NOT_OK(DoAppendRunEnd<int16_t>(run_end));
+      break;
+    case Type::INT32:
+      RETURN_NOT_OK(DoAppendRunEnd<int32_t>(run_end));
+      break;
+    case Type::INT64:
+      RETURN_NOT_OK(DoAppendRunEnd<int64_t>(run_end));
+      break;
+    default:
+      return Status::Invalid("Invalid type for run ends array: ", type_->run_end_type());
+  }
+  return Status::OK();
+}
+
+Status RunEndEncodedBuilder::CloseRun(const std::shared_ptr<const Scalar>& value,
+                                      int64_t run_length) {
+  // TODO(felipecrv): gracefully fragment runs bigger than INT32_MAX
+  if (ARROW_PREDICT_FALSE(run_length > std::numeric_limits<int32_t>::max())) {
```

Comment:
Shouldn't this depend on the offset type?

Reply:
This depends on us deciding if we want to produce huge runs. I think limiting the size of runs (not the length of logical arrays of course) is going to prevent lots of multi-language integration pains at a very low storage/memory cost. ~2B elements runs compressed to a single value is great.

Reply:
Hmm, fair enough, as long as it's documented.

Reply:
I will create an issue to open the discussion up to more people.

Reply:
Thanks for reminding me. Added a commit with it.

Reply:
Done https://github.com/apache/arrow/issues/34237